src/macros.rs

```diff
@@ -53,204 +51,121 @@ macro_rules! libc_bitflags {
-/// libc_enum!{
-///     pub enum ProtFlags {
+/// libc_enum! {
+///     pub enum ProtFlags: c_int {
 ///         PROT_NONE,
 ///         PROT_READ,
 ///         PROT_WRITE,
 ///         PROT_EXEC,
+///         /// Documentation before cfg attribute.
 ///         #[cfg(any(target_os = "linux", target_os = "android"))]
 ///         PROT_GROWSDOWN,
 ///         #[cfg(any(target_os = "linux", target_os = "android"))]
 ///         PROT_GROWSUP,
 ///     }
 /// }
+///
+/// let flag: c_int = ProtFlags::PROT_NONE.into();
+/// let flag: ProtFlags = ProtFlags::try_from(::libc::PROT_NONE).unwrap();
 /// ```
 macro_rules! libc_enum {
-    // (non-pub) Exit rule.
-    (@make_enum
-        {
-            name: $BitFlags:ident,
-            attrs: [$($attrs:tt)*],
-            entries: [$($entries:tt)*],
-        }
-    ) => {
-        $($attrs)*
-        #[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
-        enum $BitFlags {
-            $($entries)*
-        }
-    };
-
-    // (pub) Exit rule.
-    (@make_enum
-        {
-            pub,
-            name: $BitFlags:ident,
-            attrs: [$($attrs:tt)*],
-            entries: [$($entries:tt)*],
-        }
-    ) => {
-        $($attrs)*
-        #[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
-        pub enum $BitFlags {
-            $($entries)*
-        }
-    };
-
-    // (non-pub) Done accumulating.
-    (@accumulate_entries
-        {
-            name: $BitFlags:ident,
-            attrs: $attrs:tt,
-        },
-        $entries:tt;
+    // pub
+    (
+        $(#[$enum_attr:meta])*
+        pub enum $($def:tt)*
     ) => {
         libc_enum! {
-            @make_enum
-            {
-                name: $BitFlags,
-                attrs: $attrs,
-                entries: $entries,
-            }
+            @(pub)
+            $(#[$enum_attr])*
+            enum $($def)*
         }
     };
 
-    // (pub) Done accumulating.
-    (@accumulate_entries
-        {
-            pub,
-            name: $BitFlags:ident,
-            attrs: $attrs:tt,
-        },
-        $entries:tt;
+    // non-pub
+    (
+        $(#[$enum_attr:meta])*
+        enum $($def:tt)*
     ) => {
         libc_enum! {
-            @make_enum
-            {
-                pub,
-                name: $BitFlags,
-                attrs: $attrs,
-                entries: $entries,
-            }
+            @()
+            $(#[$enum_attr])*
+            enum $($def)*
         }
     };
 
-    // Munch an attr.
-    (@accumulate_entries
-        $prefix:tt,
-        [$($entries:tt)*];
-        #[$attr:meta] $($tail:tt)*
-    ) => {
-        libc_enum! {
-            @accumulate_entries
-            $prefix,
-            [
-                $($entries)*
-                #[$attr]
-            ];
-            $($tail)*
+    (
+        @($($vis:tt)*)
+        $(#[$enum_attr:meta])*
+        enum $enum:ident : $prim:ty {
+            $(
+                $(#[doc = $var_doc:tt])*
```

Comment:
The compiler emits a warning for "unused doc comment" when it sees a doc comment annotating a match arm, so I match them separately here so that I can use the `cfg` attributes on their own later.